src/source-af-xdp.c

```diff
@@ -0,0 +1,944 @@
+        fclose(fp);
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+    fclose(fp);
+
+    SCReturnInt(TM_ECODE_OK);
+}
+
+static TmEcode ConfigureBusyPolling(AFXDPThreadVars *ptv)
+{
+    const int fd = xsk_socket__fd(ptv->xsk.xsk);
+    int sock_opt = 1;
+
+    if (!ptv->xsk.enable_busy_poll) {
+        SCReturnInt(TM_ECODE_OK);
+    }
+
+    /* Kernel version must be >= 5.11 to avail of SO_PREFER_BUSY_POLL
+     * see linux commit: 7fd3253a7de6a317a0683f83739479fb880bffc8
+     */
+    if (!SCKernelVersionIsAtLeast(5, 11)) {
+        SCLogWarning(SC_WARN_AFXDP_CONF, "Kernel version older than required: v5.11,"
+                                         " upgrade kernel version to use this option.");
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    if (WriteLinuxTunables(ptv) != TM_ECODE_OK) {
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    if (setsockopt(fd, SOL_SOCKET, SO_PREFER_BUSY_POLL, (void *)&sock_opt, sizeof(sock_opt)) < 0) {
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    sock_opt = ptv->xsk.busy_poll_time;
+    if (setsockopt(fd, SOL_SOCKET, SO_BUSY_POLL, (void *)&sock_opt, sizeof(sock_opt)) < 0) {
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    sock_opt = ptv->xsk.busy_poll_budget;
+    if (setsockopt(fd, SOL_SOCKET, SO_BUSY_POLL_BUDGET, (void *)&sock_opt, sizeof(sock_opt)) < 0) {
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    SCReturnInt(TM_ECODE_OK);
+}
+
+static void AFXDPSwitchState(AFXDPThreadVars *ptv, int state)
+{
+    ptv->afxdp_state = state;
+}
+
+static TmEcode OpenXSKSocket(AFXDPThreadVars *ptv)
+{
+    int ret;
+
+    SCMutexLock(&xsk_protect.queue_protect);
+
+    if (AFXDPAssignQueueID(ptv) != TM_ECODE_OK) {
+        SCLogError(SC_ERR_SOCKET, "Failed to assign queue ID");
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    if ((ret = xsk_socket__create(&ptv->xsk.xsk, ptv->livedev->dev, ptv->xsk.queue.queue_num,
+                 ptv->umem.umem, &ptv->xsk.rx, &ptv->xsk.tx, &ptv->xsk.cfg))) {
+        SCLogError(SC_ERR_SOCKET, "Failed to create socket: %s", strerror(-ret));
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+    SCLogDebug("bind to %s on queue %u", ptv->iface, ptv->xsk.queue.queue_num);
+
+    /* For polling and socket options */
+    ptv->xsk.fd.fd = xsk_socket__fd(ptv->xsk.xsk);
+    ptv->xsk.fd.events = POLLIN;
+
+    /* Set state */
+    AFXDPSwitchState(ptv, AFXDP_STATE_UP);
+
+    SCMutexUnlock(&xsk_protect.queue_protect);
+    SCReturnInt(TM_ECODE_OK);
+}
+
+static void AFXDPCloseSocket(AFXDPThreadVars *ptv)
+{
+    if (ptv->xsk.xsk) {
+        xsk_socket__delete(ptv->xsk.xsk);
+        ptv->xsk.xsk = NULL;
+    }
+
+    if (ptv->umem.umem) {
+        xsk_umem__delete(ptv->umem.umem);
+        ptv->umem.umem = NULL;
+    }
+
+    memset(&ptv->umem.fq, 0, sizeof(struct xsk_ring_prod));
+    memset(&ptv->umem.cq, 0, sizeof(struct xsk_ring_cons));
+}
+
+static int AFXDPGetDevFlags(int fd, const char *ifname)
```

Comment:
us `GetIfaceFlags` instead? Bonus points for fixing `source-af-packet.c` as well ;)

Reply:
Not sure how I missed this... thank you 😊